app/models/prospective_employee.rb

```diff
@@ -0,0 +1,45 @@
+# frozen_string_literal: true
+
+class ProspectiveEmployee < ApplicationRecord
+  include PgSearch
+  include DynamicSearch
+  include Prospective
+
+  pg_search_scope(
+    :search,
+    against: %i[id_number staff_number first_name last_name],
+    using: { tsearch: { prefix: true } }
+  )
+
+  search_keys :unit_id, :title_id, :archived
+
+  # callbacks
+  before_create :standardization
+
+  # relations
+  belongs_to :unit
+  belongs_to :title
+
+  # validations
+  validates :date_of_birth, presence: true
+  validates :email, presence: true, length: { maximum: 255 }
+  validates :first_name, presence: true, length: { maximum: 255 }
+  validates :id_number, presence: true, uniqueness: { scope: %i[unit_id] }, length: { is: 11 }
+  validates :last_name, presence: true, length: { maximum: 255 }
+  validates :staff_number, presence: true, length: { maximum: 255 }
+  validates_with EmailAddress::ActiveRecordValidator, field: :email
+
+  # delegates
+  delegate :name, to: :title, prefix: true
+
+  def build_concrete_user(user)
+    Employee.new(user: user, title_id: title_id, staff_number: staff_number)
+  end
```

Comment:
Burada DRY olmak ugruna pek cok sey feda edilmis. Sirf User ve Employee ortak servis kullanildigi icin iki durum ortaya cikmis:

1- Farkli isler yapan ancak isimleri ayni olan 2 method?!
2- Yaptigi isle isminin alakasi olmayan methodlar. Ornegin ismi `build_concrete_user` oldugu halde `Employee` build eden ve yine ayni sekilde `Student` build eden bir method olmus.

Reply:
Rails'te `build`, `new` metodunun alias'i. `new` metodunun dokumanina bakarsak yaptigi seyin `initialization` oldugunu gorebiliriz. Dolayisiyla `build_concrete_*` gibi bir isim yerine `initialize_student` ve `initialize_employee` cok daha anlasilir.

Reply:
> 1- Farkli isler yapan ancak isimleri ayni olan 2 method?!

[With (all due) respect] bu yorumuna katılamayacağım.  Polimorfik bir metod bu: aynı isim altında farklı işler yapan metodlar.  Hatta yorumunda geçen "farklı işler yapan ancak isimleri aynı olan metodlar" cümlesi polymorphism'in sözlük girdisi gibi bir cümle. :-)

Benim burada gördüğüm problem şu.  `{Employee,Student}` var, `Prospective{Employee,Student}` var.  İkincisinden ilkine geçiş nasıl olacak, nerede olacak?  Bir önceki gerçeklemede bu `extend` yoluyla `ProspectiveService`'leriyle yapılmıştı.  Oradaki temel itirazım runtime extend gibi görünmekle birlikte, asıl sorun Employee, Student ayrımının servislere kadar inmesiydi (ProspectiveEmployee ve ProspectiveStudent için iki ayrı servis modülünün olması durumu).  Halbuki buna gerek yok.  ProspectiveService'in görevi basit tutulmalı.  ProspektifBilmemne'den Bilmemne'ye geçişi yöneten bir modül olmalı bu.  Hal böyle olunca geçiş lojiğindeki farklılıklar zaten hazır elde bulunan `ProspectiveEmployee` ve `ProspectiveStudent` nesnelerinde yönetilebilir.  Nasıl?  Bu iki nesne ProspectiveService'in kullanacağı bir interface sunar: yani (isimlendirme bir yana) `build_concrete_user`.  Bunlar çok standart hareketler.  Misal Go'da yazacak olsam da aynı şekilde yazardım.

İsim konusuna bir şey demiyorum, önceki review'de ilk aklıma gelen isim olduğundan `concrete` demiştim, değiştirilebilir.  İsimlendirmedeki espri ProspectiveService üreteceği entiteye Employee olsun Student olsun ne diyecek?  İkisini birden (polimorfik olarak) temsil edecek bir kelime gerekiyor.  Üçgen olsun, dikdörtgen olsun tüm geometrik şekillerin polimorfik bir arayüzde `shape` olarak anılması gibi.